### PR TITLE
JunOS Security Policies Module

### PR DESCRIPTION
**Pull request by pranav-bhatt**
Signed-off-by: Pranav Bhatt <adpranavb2000@gmail.com>

##### SUMMARY
This PR adds a module to handle security policies on JunOS devices.
Depends-On: https://github.com/ansible-collections/junipernetworks.junos/pull/283
##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME

junos_security_policies

##### ADDITIONAL INFORMATION

Tested against JunOS vSRX 18.4R1

### CODE REVIEW

File: LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
```
Comment (rohitthakur2590) [inline LICENSE.txt:1]: We don't need this file

File: docs/junipernetworks.junos.junos_security_policies.rst [hunk 2]
```diff
@@ -0,0 +1,1551 @@
+.. _junipernetworks.junos.junos_security_policies_module:
```
Comment (rohitthakur2590) [inline docs/junipernetworks.junos.junos_security_policies.rst:1]: This need to be updated for junos_security_policies , you can use Collection prep engine for this https://github.com/ansible-network/collection_prep

File: plugins/module_utils/network/junos/argspec/security_policies/security_policies.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+#
+# Changes should be made in the model used to
+#   generate this file or in the resource module
+#   builder template.
+#
+#############################################
+
+"""
+The arg spec for the junos_security_policies module
+"""
```
Comment (rohitthakur2590) [inline plugins/module_utils/network/junos/argspec/security_policies/security_policies.py:27]: Please add future imports: 
```
from __future__ import absolute_import, division, print_function

__metaclass__ = type
```

File: plugins/module_utils/network/junos/argspec/security_policies/security_policies.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+
+class Security_policiesArgs(object):  # pylint: disable=R0903
+    """The arg spec for the junos_security_policies module
+    """
+
+    def __init__(self, **kwargs):
+        pass
+
+    argument_spec = {'config': {'options': {'from_zones': {'elements': 'dict',
```
Comment (rohitthakur2590) [inline plugins/module_utils/network/junos/argspec/security_policies/security_policies.py:37]: need formatting here , you can run ```tox -eblack```

File: plugins/module_utils/network/junos/facts/facts.py [hunk 5]
```diff
@@ -140,9 +144,7 @@ class Facts(FactsBase):
     def __init__(self, module):
         super(Facts, self).__init__(module)
 
-    def get_facts(
```
Comment (rohitthakur2590) [inline plugins/module_utils/network/junos/facts/facts.py:143]: we don;t need to change the formatting here

File: plugins/module_utils/network/junos/facts/facts.py [hunk 6]
```diff
@@ -168,14 +170,10 @@ def get_facts(
                             "It can be installed using `pip install junos-eznc`"
                         ]
                     )
-                self.ansible_facts["ansible_net_gather_subset"].append(
```
Comment (rohitthakur2590) [inline plugins/module_utils/network/junos/facts/facts.py:171]: same as above

File: plugins/module_utils/network/junos/facts/facts.py [hunk 7]
```diff
@@ -168,14 +170,10 @@ def get_facts(
...
-                    "ofacts"
-                )
+                self.ansible_facts["ansible_net_gather_subset"].append("ofacts")
                 legacy_facts_type.remove("ofacts")
 
         if self.VALID_LEGACY_GATHER_SUBSETS:
-            self.get_network_legacy_facts(
```
Comment (rohitthakur2590) [inline plugins/module_utils/network/junos/facts/facts.py:177]: not required

File: plugins/modules/junos_ntp_global.py [hunk 8]
```diff
@@ -29,6 +29,11 @@
 
 __metaclass__ = type
 
+import debugpy
```
Comment (rohitthakur2590) [inline plugins/modules/junos_ntp_global.py:32]: to be removed

File: plugins/module_utils/network/junos/facts/security_policies/security_policies.py [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+# -*- coding: utf-8 -*-
+# Copyright 2022 Red Hat
+# GNU General Public License v3.0+
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""
+The junos security_policies fact class
+It is in this file the configuration is collected from the device
+for a given resource, parsed, and the facts tree is populated
+based on the configuration.
+"""
```
Comment (rohitthakur2590) [inline plugins/module_utils/network/junos/facts/security_policies/security_policies.py:11]: need to add future imports

File: plugins/modules/junos_security_policies.py [hunk 10]
```diff
@@ -0,0 +1,662 @@
...
+    - merged
+    - replaced
+    - overridden
+    - deleted
+    - rendered
+    - gathered
+    - parsed
+    default: merged
+    description:
+      - The state the configuration should be left in
```
Comment (rohitthakur2590) [inline plugins/modules/junos_security_policies.py:544]: we can add note for replaced and overridden behaviour as well

File: plugins/modules/junos_security_policies.py [hunk 11]
```diff
@@ -0,0 +1,662 @@
...
+      - The state I(parsed) reads the configuration from C(running_config) option and
+        transforms it into JSON format as per the resource module parameters and the
+        value is returned in the I(parsed) key within the result The value of C(running_config)
+        option should be the same format as the output of command
+        I(show security policies detail) executed on device For state I(parsed) active
+        connection to remote host is not required
+    type: str
+"""
+EXAMPLES = """
+
```
Comment (rohitthakur2590) [inline plugins/modules/junos_security_policies.py:563]: Update with examples

File: plugins/modules/junos_security_policies.py [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+commands:
+  description: The set of commands pushed to the remote device.
+  returned: when state is I(merged), I(replaced), I(overridden), I(deleted) or I(purged)
+  type: list
+  sample:
+    - "logging console 3"
+    - "logging monitor 4"
+    - "logging ip access-list cache entries 16384"
+    - "logging ip access-list cache interval 200"
+    - "logging ip access-list cache threshold 5000"
```
Comment (rohitthakur2590) [inline plugins/modules/junos_security_policies.py:609]: this should be xml content

File: plugins/modules/junos_security_policies.py [hunk 13]
```diff
@@ -0,0 +1,662 @@
...
+rendered:
+  description: The provided configuration in the task rendered in device-native format (offline).
+  returned: when state is I(rendered)
+  type: list
+  sample:
+    - "logging ip access-list cache entries 4096"
```
Comment (rohitthakur2590) [inline plugins/modules/junos_security_policies.py:615]: same as above

File: plugins/modules/junos_security_policies.py [hunk 14]
```diff
@@ -0,0 +1,662 @@
...
+    - "no logging ip access-list cache interval 200"
+    - "no logging ip access-list cache threshold 5000"
+    - "no logging origin-id hostname"
+    - "logging origin-id ip 192.0.2.100"
+    - "logging server 198.51.100.101 1 port 6538 use-vrf management"
+gathered:
+  description: Facts about the network resource gathered from the remote device as structured data.
+  returned: when state is I(gathered)
+  type: list
```
Comment (rohitthakur2590) [inline plugins/modules/junos_security_policies.py:624]: It should be dict as per argspec

File: plugins/modules/junos_security_policies.py [hunk 15]
```diff
@@ -0,0 +1,662 @@
...
+  sample: >
+    This output will always be in the same format as the
+    module argspec.
+parsed:
+  description: The device native config provided in I(running_config) option parsed into structured data as per module argspec.
+  returned: when state is I(parsed)
+  type: list
```
Comment (rohitthakur2590) [inline plugins/modules/junos_security_policies.py:631]: should be dict

File: plugins/modules/junos_security_policies.py [hunk 16]
```diff
@@ -0,0 +1,662 @@
...
+)
+
+
+def main():
+    """
+    Main entry point for module execution
+
+    :returns: the result form module invocation
+    """
+    module = AnsibleModule(
```
Comment (rohitthakur2590) [inline plugins/modules/junos_security_policies.py:653]: you can add empty config checks here

File: tests/integration/network-integration.cfg [hunk 17]
```diff
@@ -1,4 +0,0 @@
-[persistent_connection]
```
Comment (rohitthakur2590) [inline tests/integration/network-integration.cfg:1]: this is not required

File: tests/integration/targets/junos_security_policies/tests/netconf/backups/empty_config.yaml [hunk 18]
```diff
@@ -0,0 +1,49 @@
...
+  ignore_errors: true
+  junipernetworks.junos.junos_security_policies:
+    config:
+    state: merged
+
+- assert:
+    that:
+      - result.msg == 'value of config parameter must not be empty for state merged'
+
+- name: Replaced with empty config should give appropriate error message
```
Comment (rohitthakur2590) [inline tests/integration/targets/junos_security_policies/tests/netconf/backups/empty_config.yaml:17]: We don't need this file

**Review by rohitthakur2590 — CHANGES_REQUESTED** [review]

File: plugins/module_utils/network/junos/facts/facts.py (hunk 5 above)
Comment (pranav-bhatt) [reply plugins/module_utils/network/junos/facts/facts.py:143]: the formatting changes are introduced due to black. The default line length is 88 characters. Is that not the default formatter used?

File: tests/integration/network-integration.cfg (hunk 17 above)
Comment (pranav-bhatt) [reply tests/integration/network-integration.cfg:1]: are you saying that this file isn't required? Because its already removed

File: README.md [hunk 19]
```diff
@@ -12,6 +12,8 @@ This collection has been tested against Juniper Networks Junos OS 18.4R1.
 
 This collection has been tested against following Ansible versions: **>=2.9.10**.
 
+For collections that support Ansible 2.9, please ensure you update your `network_os` to use the 
+fully qualified collection name (for example, `cisco.ios.ios`).
```
Comment (rohitthakur2590) [inline README.md:16]: junos reference wold be better

File: meta/runtime.yml [hunk 20]
```diff
@@ -98,6 +98,10 @@ plugin_routing:
       redirect: junipernetworks.junos.junos
     ntp_global:
       redirect: junipernetworks.junos.junos
+    junos_security_policies:
+      redirect: junipernetworks.junos.junos
+    security_policies:
+      redirect: junipernetworks.junos.junos
```
Comment (rohitthakur2590) [inline meta/runtime.yml:104]: module redirection also needs to be updated

**Review by rohitthakur2590 — CHANGES_REQUESTED** [review]

File: tests/unit/modules/network/junos/test_junos_security_policies.py [hunk 21]
```diff
@@ -0,0 +1,626 @@
...
+        self.load_config = self.mock_load_config.start()
+
+        self.mock_commit_configuration = patch(
+            "ansible_collections.junipernetworks.junos.plugins.module_utils.network.junos.config.security_policies.security_policies.commit_configuration"
+        )
+        self.mock_commit_configuration = self.mock_commit_configuration.start()
+
+        self.mock_execute_show_command = patch(
+            "ansible_collections.junipernetworks.junos.plugins.module_utils.network.junos.facts.security_policies.security_policies."
+            "security_policiesFacts.get_device_data"
```
Comment (NilashishC) [inline tests/unit/modules/network/junos/test_junos_security_policies.py:66]:
```suggestion
            "Security_policiesFacts.get_device_data"
```
```suggestion
            "security_policiesFacts.get_device_data"
```

File: README.md (hunk 19 above)
Comment (trishnaguha) [reply README.md:15]: Should this be junipernetworks.junos.junos?

**Comment by pranav-bhatt** [comment]
recheck

**Comment by pranav-bhatt** [comment]
recheck

**Review by rohitthakur2590 — APPROVED** [review]

**Review by justjais — APPROVED** [review]

**Comment by pranav-bhatt** [comment]
recheck

**Comment by rohitthakur2590** [comment]
recheck

**Pull request merged**